File: cpp/sublist/sublist.cpp
#include "sublist.h"
#include <algorithm>

namespace sublist {

// Helper function to check if list_one is a subsequence of list_two
bool is_subsequence(const std::vector<int>& list_one, const std::vector<int>& list_two) {
    if (list_one.empty()) {
        return true; // Empty list is a subsequence of any list
    }
    
    if (list_one.size() > list_two.size()) {
        return false; // First list can't be a subsequence if it's longer
    }
    
    // Check for subsequence at each possible starting position in list_two
    for (size_t i = 0; i <= list_two.size() - list_one.size(); ++i) {
        bool found = true;
        for (size_t j = 0; j < list_one.size(); ++j) {
            if (list_one[j] != list_two[i + j]) {
                found = false;
                break;
            }
        }
        if (found) {
            return true;
        }
    }
    
    return false;
}

List_comparison sublist(const std::vector<int>& list_one, const std::vector<int>& list_two) {
    // Check for equality
    if (list_one == list_two) {
        return List_comparison::equal;
    }
    
    // Check if list_one is a sublist of list_two
    if (is_subsequence(list_one, list_two)) {
        return List_comparison::sublist;
    }
    
    // Check if list_two is a sublist of list_one (i.e., list_one is a superlist)
    if (is_subsequence(list_two, list_one)) {
        return List_comparison::superlist;
    }
    
    // If none of the above, the lists are unequal
    return List_comparison::unequal;
}

}  // namespace sublist

File: go/sublist/sublist.go
package sublist

// Relation type is defined in relations.go file.

func Sublist(l1, l2 []int) Relation {
	// If both lists are empty, they are equal
	if len(l1) == 0 && len(l2) == 0 {
		return RelationEqual
	}

	// If l1 is empty, it's a sublist of l2
	if len(l1) == 0 {
		return RelationSublist
	}

	// If l2 is empty, l1 is a superlist of l2
	if len(l2) == 0 {
		return RelationSuperlist
	}

	// If both lists have the same length, check if they are equal
	if len(l1) == len(l2) {
		equal := true
		for i := range l1 {
			if l1[i] != l2[i] {
				equal = false
				break
			}
		}
		if equal {
			return RelationEqual
		}
	}

	// Check if l1 is a sublist of l2
	if len(l1) < len(l2) && isSubsequence(l1, l2) {
		return RelationSublist
	}

	// Check if l1 is a superlist of l2
	if len(l1) > len(l2) && isSubsequence(l2, l1) {
		return RelationSuperlist
	}

	// If none of the above, they are unequal
	return RelationUnequal
}

// isSubsequence checks if list1 is a subsequence of list2
func isSubsequence(list1, list2 []int) bool {
	if len(list1) > len(list2) {
		return false
	}

	for i := 0; i <= len(list2)-len(list1); i++ {
		match := true
		for j := 0; j < len(list1); j++ {
			if list1[j] != list2[i+j] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}
